### PR TITLE
[305] Submission Details - Judging Status and Comments

### LINKED ISSUE #80: Submission Detail View for Challenge Managers
### User story
As a challenge manager, in order to manage submissions and their evaluations process efficiently, I would like to be able to view all the details of a single submissions.

**Acceptance Criteria:**

- [x] From a submissions list view, a challenge manager can click on a submission and is displayed the following submission and evaluations information:
  - [x] Submission ID
  - [x] Submission and public solvers information (materials a public solver submitted as part of the entry):
    - [x] Brief description
    - [x] Description
    - [x] #284 
    - [x] External URL
    - [x] Submission status
    - [x] Submitter email
    - [x] Last updated date
- [x] A challenge manager has an option to either hide (collapse) submissions materials or open submission materials on a new tab
  - #304 
- [x] A challenge manager can select a submission for evaluation (formerly "pre-screen"), #81 
  - [x] If a submission is eligible for evaluation, a submission evaluation section is displayed, #85 
- [x] A challenge manager can select a submission for advancement, #84 
- [x] A challenge manager can enter and save notes for record keeping 
  - [x] Freeform input field is not required and is only saved and visible for challenge managers

**Development tasks:**

- [x] #284 
- [x] #345 
- [ ] ~#351~ 

### Definition of Done

**Doing (dev team)**
- [x] Code complete 
- [x] Code is organized appropriately
- [x] Any known trade offs are documented in the associated GH issue
- [x] Code is documented, modules, shared functions, etc.
- [x] Automated testing has been added or updated in response to changes in this PR
- [x] The feature is smoke tested to confirm it meets requirements
- [x] Database changes have been peer reviewed for index changes and performance bottlenecks
- [x] PR that changes or adds UI
  - [x] include a screenshot of the WAVE report for the altered pages
  - [x] Confirm changes were validated for mobile responsiveness
- [x] PR approved / Peer reviewed 
- [x] Security scans passed
- [x] Automate accessibility tests passed
- [x] Build process and deployment is automated and repeatable
- [x] Feature toggles if appropriate
- [x] Deploy to staging
- [x] Move card to testing column in the board
---
**Staging**
- [x] Accessibility tested (Marni)
  - [x]  Keyboard navigation
  - [x]  Focus confirmed
  - [x] Color contrast compliance
  - [x] Screen reader testing 
- [x] Usability testing: mobile and desktop (Tracy or Marni)
- [x] Cross browser testing - UI rendering is performant on below listed devices/browsers (Tracy or Marni)
  - [x] Windows/Chrome
  - [x] Windows/Edge
  - [x] Mac/Chrome
  - [x] Mac/Safari
  - [x] iOS/Safari
- [x] AC review (Renata)
- [ ] ~Deploy to production (production-like environment for eval capability) (dev team)~
- [x] Move to production column in the board
---
**Production**
- [ ] User and security documentation has been reviewed for necessary updates (Renata/Tracy/Dev team) 
- [ ] PO / PM approved (Jarah or Renata)
- [ ] AC is met and it works as expected (Jarah or Renata)
- [ ] Move to done column in the board (Jarah or Renata)

### LINKED ISSUE #305: Submission Details: Judging Status buttons and Comments & Notes textarea
# What/Why
Implement interactive inputs on the main details section.
- Eligible for evaluation checkbox/button
- Selected to advance checkbox/button
- Comments and Notes textarea

### Figma Link

[link-to-figma](https://www.figma.com/design/VfdBFG8ShVnOStdSNYy9fR/Challenge.Gov-Component-Library)

## Acceptance Criteria
- [ ] Update requests can use the SubmissionsController#update route - `PATCH /submissions/:id`
- [ ] Status buttons update the `judging_status` column when they are clicked
  - [ ] Eligible for evaluation updates `submission.judging_status = :selected`
  - [ ] Selected to advance updates `submission.judging_status = :winner`
- [ ] When Eligible for evaluation is selected the Submission Evaluation section is displayed (see: Out of Scope note)
- [ ] Selected to advance is _disabled & deselected_ when Eligible for evaluation is deselected
- [ ] Conversely, Eligible for evaluation is _disabled & selected_ when Selected to advance is selected
- [ ] Comments and Notes textarea updates the `submission.comments` attribute when the Save button is clicked
  - limit 3000 characters

_Disregard the following if this work **does not** include any UI changes_
- [ ] WAVE screenshot
- [ ] Automated a11y tests (axe-rspec specs)
- [ ] Keyboard accessibility (tab navigation, form interactions)
- [ ] Responsiveness at desktop and mobile

## Out of Scope
Implementing Submission Evaluation section is out of scope. Just the section header can be shown for now.

### PR DESCRIPTION
**Pull request by stonefilipczak**
Allows a challenge manager to manipulate judging status and comments. 

State 1: Submission not eligible for evaluation (neither box checked, second box disabled)
Screen Shot 2024-12-10 at 11 38 22 AM

State 2: Submission eligible for evaluation (first box checked, second box enabled)
Screen Shot 2024-12-10 at 11 39 13 AM

State 3: Submission selected to advance (both boxes checked, first box disabled)
Screen Shot 2024-12-10 at 11 39 59 AM

Screen Shot 2024-12-10 at 11 36 27 AM

### CODE REVIEW

File: spec/system/submission_details_spec.rb
Comment (stepchud) [inline spec/system/submission_details_spec.rb:1]: woohoo, I was about to request to add this spec but you are way ahead of me 🪄

**Review by stepchud — COMMENTED** [review]
Found a potential a11y issue, waiting for recommendation from the product team
[GSA Slack thread](https://gsa-tts.slack.com/archives/GTETE7W82/p1734035573400439)

File: app/views/submissions/_judging_status.html.erb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    <%= form_with(model: @submission, url: submission_path(@submission), class: "maxw-mobile-lg") do |form| %>
+        <div class="usa-checkbox" id="selected-to-advance">
+            <input
+            class="usa-checkbox__input usa-checkbox__input--tile"
+            id="judging-status-winner"
+            type="checkbox"
+            name="submission[judging_status]"
+            value="winner"
+            <%= unless @submission.judging_status.in?(["selected", "winner"]) then "disabled" end %>
+            <%= if @submission.judging_status == "winner" then "checked" end %>
```
Comment (stepchud) [inline app/views/submissions/_judging_status.html.erb]: how do you feel about the postfix version of conditionals like this?
```suggestion
            <%= "disabled" unless @submission.judging_status.in?(["selected", "winner"]) %>
            <%= "checked" if @submission.judging_status == "winner" %>
```

File: spec/system/submission_details_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    let(:challenge) { create_challenge(user: user, title: "Boston Tea Party Cleanup") }
+    let(:submission) { create(:submission, manager: user, challenge: challenge) }
+
+
+    before { system_login_user(user) }
+
+    it "submission details page is accessible" do
+      visit submission_path(submission)
+      expect(user.role).to eq("challenge_manager")
+      expect(page).to have_content(submission.id)
```
Comment (stepchud) [inline spec/system/submission_details_spec.rb]: I'd like to use more specific rules to ensure we are testing the correct content on the page, validating unique content. I added the `have_css` selector for request specs (e20849c), and I'd like to take advantage of it by checking the css as well as page content match. in this case:
```suggestion
      expect(page).to have_css('h1', text: "Submission ID #{submission.id}")
```
the other tests look good to me for page interaction specification 👍

File: spec/system/submission_details_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      expect(page).to(be_axe_clean)
+    end
+
+    it "allows manipulation of judging status" do
+      visit submission_path(submission)
+
+      find('#eligible-for-evaluation').click
```
Comment (stepchud) [inline spec/system/submission_details_spec.rb]: my linter suggests this locally for [Capybara/SpecificFinders](https://docs.rubocop.org/rubocop-capybara/cops_capybara.html#capybaraspecificfinders)
```suggestion
      find_by_id('eligible-for-evaluation').click
```

File: spec/system/submission_details_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      updated_submission = Submission.find(submission.id)
+      expect(updated_submission.judging_status).to eq('selected')
+
+      find('#selected-to-advance').click
```
Comment (stepchud) [inline spec/system/submission_details_spec.rb]: my linter suggests this locally for [Capybara/SpecificFinders](https://docs.rubocop.org/rubocop-capybara/cops_capybara.html#capybaraspecificfinders)
```suggestion
      find_by_id('selected-to-advance').click
```

**Review by stepchud — DISMISSED** [review]
linter suggestion

**Comment by stepchud** [comment]
@stonefilipczak sorry this took a bit of a different turn due to a11y concerns. the current decision is to *not submit* the form when the checkboxes are clicked or updated. I didn't realize the a11y implications when I wrote the requirements. can you switch this to more of a standard "single form" approach where the user needs to submit the form Save button action in order to persist the status changes?

File: app/views/submissions/_judging_status.html.erb (hunk 1 above)
Comment (stonefilipczak) [reply app/views/submissions/_judging_status.html.erb]: love it

**Review by stepchud — CHANGES_REQUESTED** [review]
The all-in-one updates seem to be working, however the checkbox disabled / checked states are not implemented. I'm able to select and deselect them independent of each other after the initial page load.

https://github.com/user-attachments/assets/57826dc0-8bd6-4e8a-ade4-7685a38ac4fa

**Review by stepchud — APPROVED** [review]

**Pull request merged**